File: src/java/src/test/java/fr/mcorbin/mirabelle/memtable/EngineTest.java
package fr.mcorbin.mirabelle.memtable;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;

public class EngineTest {

    @Test
    @DisplayName("Adding values to the engine, trigger several cleanups")
    public void testAddCleanup() {
        Engine engine = new Engine(60, 30);;
        HashMap<String, String> labels = new HashMap<>();
        labels.put("host", "mcorbin.fr");
        Serie serie = new Serie("cpu", labels);
        engine.add(1, serie, 1);
        assert (engine.valuesFor(serie).containsAll(Arrays.asList(1)));
        assertEquals(1, engine.valuesFor(serie).size());

        engine.add(59, serie, 2);
        assert (engine.valuesFor(serie).containsAll(Arrays.asList(2)));
        assertEquals(2, engine.valuesFor(serie).size());

        engine.add(62, serie, 3);
        assert (engine.valuesFor(serie).containsAll(Arrays.asList(2, 3)));
        assertEquals(2, engine.valuesFor(serie).size());

        engine.add(62, serie, 4);
        assertEquals(2, engine.valuesFor(serie).size());
        assert (engine.valuesFor(serie).containsAll(Arrays.asList(2, 4)));
        engine.add(70, serie, 5);
        engine.add(95, serie, 6);
        engine.add(96, serie, 7);
        engine.add(121, serie, 8);

        Collection<Object> values = engine.valuesFor(serie);
        assertEquals(3, values.size());
        assert (engine.valuesFor(serie).containsAll(Arrays.asList(6, 7, 8)));

        engine.add(300, serie, 7);
        assertEquals(3, values.size());
        assertEquals(1, engine.valuesFor(serie).size());
    }
}

File: test/mirabelle/java/engine_test.clj
(ns mirabelle.java.engine-test
  (:require [clojure.test :refer :all]
            [mirabelle.db.memtable :as memtable])
  (:import fr.mcorbin.mirabelle.memtable.Engine
           fr.mcorbin.mirabelle.memtable.Serie))

(deftest test-add-cleanup
  (let [engine ^Engine (Engine. 60 30)
        serie ^Serie (memtable/->serie "cpu" {"host" "mcorbin.fr"})
        ]
    (.add engine 1 serie 1)
    (is (= (.valuesFor engine serie) [1]))
    (.add engine 59 serie 2)
    (is (= (.valuesFor engine serie) [1 2]))

    (.add engine 62 serie 3)
    (is (= (.valuesFor engine serie) [2 3]))
    (.add engine 62 serie 4)
    (is (= (.valuesFor engine serie) [2 4]))
    (.add engine 70 serie 5)
    (.add engine 95 serie 6)
    (.add engine 96 serie 7)
    (.add engine 121 serie 8)
    (is (= (.valuesFor engine serie) [6 7 8]))

    (.add engine 300 serie 7)
    (is (= (.valuesFor engine serie) [7]))))

